weevely.1
.TH WEEVELY "1" "October 2018" "weevely 4.0.2" "User Commands"
.SH NAME
Weevely \- Weaponized web shell
.SH DESCRIPTION
.PP
A web shell designed for post\-exploitation purposes that can
be extended over the network at runtime.
.PP
Upload weevely PHP agent to a target web server to get remote shell
access to it.
Once connected you can make use of the more than 30 modules to assist
administrative tasks, maintain access, provide situational awareness,
elevate privileges, and spread into the target network.
.SH SYNOPSIS
.nf
\f[C]
Run terminal to the target
\fBweevely\ <URL>\ <password>\ [cmd]\fR

Generate backdoor agent
\fBweevely\ generate\ <password>\ <path>\fR

Load session file
\fBweevely\ session\ <path>\fR
\f[]

.SS Features
.IP \[bu] 2
Shell access to the target
.IP \[bu] 2
SQL console pivoting on the target
.IP \[bu] 2
HTTP/HTTPS proxy to browse through the target
.IP \[bu] 2
Upload and download files
.IP \[bu] 2
Spawn reverse and direct TCP shells
.IP \[bu] 2
Audit remote target security
.IP \[bu] 2
Run Meterpreter payloads
.IP \[bu] 2
Port scan pivoting on target
.IP \[bu] 2
Mount the remote filesystem
.IP \[bu] 2
Bruteforce SQL accounts pivoting on the target
.SS Agent
.PP
The agent is a small, polymorphic PHP script hardly detected by AV and
the communication protocol is obfuscated within HTTP requests.
.SS Modules
.PP
.TS
tab(@);
lw(27.4n) lw(42.6n).
T{
Module
T}@T{
Description
T}
_
T{
:audit_filesystem
T}@T{
Audit the file system for weak permissions.
T}
T{
:audit_suidsgid
T}@T{
Find files with SUID or SGID flags.
T}
T{
:audit_disablefunctionbypass
T}@T{
Bypass disable_function restrictions with mod_cgi and .htaccess.
T}
T{
:audit_etcpasswd
T}@T{
Read /etc/passwd with different techniques.
T}
T{
:audit_phpconf
T}@T{
Audit PHP configuration.
T}
T{
:shell_sh
T}@T{
Execute shell commands.
T}
T{
:shell_su
T}@T{
Execute commands with su.
T}
T{
:shell_php
T}@T{
Execute PHP commands.
T}
T{
:system_extensions
T}@T{
Collect PHP and webserver extension list.
T}
T{
:system_info
T}@T{
Collect system information.
T}
T{
:system_procs
T}@T{
List running processes.
T}
T{
:backdoor_reversetcp
T}@T{
Execute a reverse TCP shell.
T}
T{
:backdoor_tcp
T}@T{
Spawn a shell on a TCP port.
T}
T{
:backdoor_meterpreter
T}@T{
Start a meterpreter session.
T}
T{
:bruteforce_sql
T}@T{
Bruteforce SQL database.
T}
T{
:file_gzip
T}@T{
Compress or expand gzip files.
T}
T{
:file_clearlog
T}@T{
Remove string from a file.
T}
T{
:file_check
T}@T{
Get attributes and permissions of a file.
T}
T{
:file_upload
T}@T{
Upload file to remote filesystem.
T}
T{
:file_webdownload
T}@T{
Download an URL.
T}
T{
:file_tar
T}@T{
Compress or expand tar archives.
T}
T{
:file_download
T}@T{
Download file from remote filesystem.
T}
T{
:file_bzip2
T}@T{
Compress or expand bzip2 files.
T}
T{
:file_edit
T}@T{
Edit remote file on a local editor.
T}
T{
:file_grep
T}@T{
Print lines matching a pattern in multiple files.
T}
T{
:file_ls
T}@T{
List directory content.
T}
T{
:file_cp
T}@T{
Copy single file.
T}
T{
:file_rm
T}@T{
Remove remote file.
T}
T{
:file_upload2web
T}@T{
Upload file automatically to a web folder and get corresponding URL.
T}
T{
:file_zip
T}@T{
Compress or expand zip files.
T}
T{
:file_touch
T}@T{
Change file timestamp.
T}
T{
:file_find
T}@T{
Find files with given names and attributes.
T}
T{
:file_mount
T}@T{
Mount remote filesystem using HTTPfs.
T}
T{
:file_enum
T}@T{
Check existence and permissions of a list of paths.
T}
T{
:file_read
T}@T{
Read remote file from the remote filesystem.
T}
T{
:file_cd
T}@T{
Change current working directory.
T}
T{
:sql_console
T}@T{
Execute SQL query or run console.
T}
T{
:sql_dump
T}@T{
Multi dbms mysqldump replacement.
T}
T{
:net_mail
T}@T{
Send mail.
T}
T{
:net_phpproxy
T}@T{
Install PHP proxy on the target.
T}
T{
:net_curl
T}@T{
Perform a curl\-like HTTP request.
T}
T{
:net_proxy
T}@T{
Run local proxy to pivot HTTP/HTTPS browsing through the target.
T}
T{
:net_scan
T}@T{
TCP Port scan.
T}
T{
:net_ifconfig
T}@T{
Get network interfaces addresses.
T}
.TE
.SH SEE ALSO
https://github.com/epinna/weevely3/wiki
.SH AUTHOR
weevely is developed by The Weevely Developers, this manpage was made by Emilio <epinna> and Samuel Henrique <samueloph@debian.org> based on weevely's README.md and can be used by other projects as well.



CHANGELOG.md
# Change Log

## [v4.0.2] - 

### Fixed
- Broken file_edit #133

## [v4.0.1] - 2020-01-06
### Removed
- Remove PHP minification

### Fixed
- Broken cd #122

## [v4.0.0] - 2019-12-26
### Added
- Full port to Python 3

### Fixed
- Module net_phpproxy
- Fixes alias management #117

### Removed
- Old backdoor formats LegacyCookie, LegacyReferrer, and Stegaref
- Module backdoor_meterpreter

## [v3.7.0] - 2018-10-15
### Fixed
- Fix vector handling in audit_etcpasswd #93

### Added
- HTTPS proxy support
- Support OPTIONS request in net_curl module
- Use httpbin for net_proxy testing

## [v3.6.2] - 2018-06-27
### Fixed
- Remove audit_linuxprivchecker module

## [v3.6.1] - 2018-06-26
### Fixed
- Fixed corrupted session file #83
- Vendor files licensing

### Added
- Man page

## [v3.6] - 2018-06-02
### Fixed
- PHP 7 support
- Add exceptions catches

### Added
- ObfPost is the default channel to obfuscate traffic in POST requests
- Travic-CI integration

## [v3.5] - 2017-23-11
### Fixed
- Connection to HTTPS sites with wrong certificates
- Fix net_phpproxy params
- Fix broken audit_etcpasswd
- Fix strings passing on SQL console

### Added
- Dockerized test suite

## [v3.4] - 2016-11-02
### Fixed
- Stored/Passed/Default arguments handling

### Added
- Backdoor_meterpreter module
- System_procs module by @paddlesteamer
- Support for database/schema selection for PostgreSQL by @caeaguilar

## [v3.3.1] - 2016-05-12
### Fixed
- File grep module wrong grepping folders
- 500 error testing pcntl_fork in shell sh module

### Added
- Clearlog and Mail module by @AppoPro
- Socat vector in backdoor tcp module
- Legacycookie generator
- Upload2web module content and simulate options
- Disable_function bypass module

## [v3.2.0] - 2015-06-29
### Fixed
- Basic Windows support
- OS X support
- Improve Stegaref channel referrer templates
- Reset on channel session variables changes

### Added
- Python requirements.txt
- Encoding support for sql_console
- Output redirection and inverse grep for file_grep
- Run actions on start depending from the session load
- Proxy and SOCKS support
- Unset session variables
- Show session variables

## [v3.1.0] - 2015-04-07
### Added
- Module bruteforce_sql
- Additional HTTP headers
- Direct command execution from weevely.py argument
- Whoami, hostname, pwd, uname aliases
- Module file_cp
- CHANGELOG

## v3.0.0 - 2015-01-17
### Added
- Module sql_dump
- Module sql_console
- Module bruteforce_sqlusers
- Module bruteforce_sql
- Module file_mount
- Module file_enum
- Module file_download
- Module file_touch
- Module file_rm
- Module file_edit
- Module file_read
- Module file_upload
- Module file_upload2web
- Module backdoor_reversetcp
- Module backdoor_tcp
- Module audit_suidsgid
- Module file_find
- Module audit_filesystem
- Module audit_phpconf
- Module audit_etcpasswd
- Module net_phpproxy
- Module net_ifconfig
- Module net_proxy
- Module net_scan
- Module file_grep
- Module net_curl
- Modules file_zip, file_tar, file_gzip, and file_bzip2
- Support of legacy LegacyCookie and LegacyReferrer channels

### Removed
- Module audit_userfiles
- Module audit_mapwebfiles


[unreleased]: https://github.com/epinna/weevely3/commit/HEAD
[v3.1.0]: https://github.com/epinna/weevely3/releases/tag/v3.1.0
[v3.2.0]: https://github.com/epinna/weevely3/releases/tag/v3.2.0
[v3.3.1]: https://github.com/epinna/weevely3/releases/tag/v3.3.1
[v3.4]: https://github.com/epinna/weevely3/releases/tag/v3.4
[v3.5]: https://github.com/epinna/weevely3/releases/tag/v3.5
[v3.6]: https://github.com/epinna/weevely3/releases/tag/v3.6
[v3.6.1]: https://github.com/epinna/weevely3/releases/tag/v3.6.1
[v3.6.2]: https://github.com/epinna/weevely3/releases/tag/v3.6.2
[v3.7.0]: https://github.com/epinna/weevely3/releases/tag/v3.7.0
[v4.0.0]: https://github.com/epinna/weevely3/releases/tag/v4.0.0
[v4.0.1]: https://github.com/epinna/weevely3/releases/tag/v4.0.1


weevely.py
#!/usr/bin/env python3
import glob
import os
import pprint
import sys

from core import generate
from core import messages
from core import modules
from core.argparsers import CliParser
from core.config import agent_templates_folder_path, obfuscators_templates_folder_path
from core.loggers import log, dlog
from core.sessions import SessionURL, SessionFile
from core.terminal import Terminal
from core.weexceptions import FatalException, ArgparseError

if sys.stdout.encoding is None:
    print("Please set PYTHONIOENCODING=UTF-8 running 'export PYTHONIOENCODING=UTF-8' before starting Weevely.")
    exit(1)


def main(arguments):

    if arguments.command == 'generate':

        obfuscated = generate.generate(
            password = arguments.password,
            obfuscator = arguments.obfuscator,
            agent = arguments.agent
        )

        generate.save_generated(obfuscated, arguments.path)

        if arguments.path != '-':
            log.info(messages.generate.generated_backdoor_with_password_s_in_s_size_i %
                     (arguments.path,
                      arguments.password,
                      len(obfuscated)
                      )
                     )

        return

    elif arguments.command == 'terminal':
        session = SessionURL(
            url = arguments.url,
            password = arguments.password
        )

    elif arguments.command == 'session':
        session = SessionFile(arguments.path)

    dlog.debug(
        pprint.pformat(session)
    )

    modules.load_modules(session)

    if not arguments.cmd:
        Terminal(session).cmdloop()
    else:
        term = Terminal(session)
        term.precmd(arguments.cmd)
        term.onecmd(arguments.cmd)

if __name__ == '__main__':

    parser = CliParser(prog='weevely')
    subparsers = parser.add_subparsers(dest = 'command')

    terminalparser = subparsers.add_parser('terminal', help='Run terminal or command on the target')
    terminalparser.add_argument('url', help = 'The agent URL')
    terminalparser.add_argument('password', help = 'The agent password')
    terminalparser.add_argument('cmd', help = 'Command', nargs='?')

    sessionparser = subparsers.add_parser('session', help='Recover an existing session')
    sessionparser.add_argument('path', help = 'Session file path')
    sessionparser.add_argument('cmd', help = 'Command', nargs='?')

    agents_available = [
        os.path.split(agent)[1].split('.')[0] for agent in
        glob.glob('%s/*.tpl' % agent_templates_folder_path)
    ]

    obfuscators_available = [
        os.path.split(agent)[1].split('.')[0] for agent in
        glob.glob('%s/*.tpl' % obfuscators_templates_folder_path)
    ]

    generateparser = subparsers.add_parser('generate', help='Generate new agent')
    generateparser.add_argument('password', help = 'Agent password')
    generateparser.add_argument('path', help = 'Agent file path')
    generateparser.add_argument(
        '-obfuscator', #The obfuscation method
        choices = obfuscators_available,
        default = 'phar'
        )
    generateparser.add_argument(
        '-agent', #The agent channel type
        choices = agents_available,
        default = 'obfpost_php'
        )

    parser.set_default_subparser('terminal')

    try:
        arguments = parser.parse_args()
    except ArgparseError:
        parser.exit()

    try:
        main(arguments)
    except (KeyboardInterrupt, EOFError):
        log.info('Exiting.')
    except FatalException as e:
        log.critical('Exiting: %s' % e)
